### ISSUE TITLE
Support for DSM 5.X

### ISSUE THREAD

**Issue by meoso** [zerotier/ZeroTierNAS#12]
https://download.zerotier.com/dist/zerotier-1.2.4-syn-88f6281-6.1.spk is offered, but non existent.

Maybe because the synology OS is stuck at 5.x and 6.x is not available?  I'm still using ZerTier-One 1.1.14-4 and it technically works. (http://pkg.entware.net/binaries/armv5/Packages.html)

**Comment by lucktu** [comment]
I want too.

**Comment by joseph-henry** [comment]
Working to fix this. At the worst case I'll just re-package one of the 1.2.4 binaries I cross compiled a while back for this arch. Can you both tell me the exact models and exact DSM versions? 

Thanks.

**Comment by joseph-henry** [comment]
Just added packages for both `88F6281` and `88F6282`. I've only tested `88F6282`. So the link above should work now.

Note, this was built for DSM 6.1 as my test device is on that version. If you need it to run this on DSM 5.X you can extract the binary from the package and manually install it. It should be compatible.

**Comment by lucktu** [comment]
Joseph Henry: Hi!

          It's very nice to see you again!!


      My NAS is DS213j (Marvell Armada 370 88F6707), dsm is DSM 5.2-5967 Update 6. I want zerotierone for it (dsm 5.2).


      I test the zerotier-1.2.4-syn-88f6281-6.1.spk in ds212J (Yes,it's 212J, DSM 5.2-5967 Update 6), but I'm sorry, it's bad. in dsm, it's stopped for ever.

      In addition, zerotier-one.spk is ok in ds212J, only in dsm, it's OFFLINE for ever, but go to ssh, run zerotier-cli join XXXX, it's ok, but it's mortifying, reboot, it's bad again.

      In addition, dsm is demoted by manual. (It's danger!! referenc: http://www.hkepc.com/forum/viewthread.php?tid=1652223 ) , 1. down two older dsm(http://usdl.synology.com/download/DSM/)(exemple:5.2-5565 & 5.2-5967 ), unzip it (5.2-5565), get VERSION form it. 2.(in ssh) rename /etc.defaults/VERSION to VERSION0. 3.up the old version to here. 4.return to dsm, update. I test: 4.0-2636 to 2233 (by 2228), it's ok. Maybe, referenc is overdue.

      How can I help you? It's for me truly ^_^



      Thanks a lot!! Best regards for you!

________________________________
发件人: Joseph Henry <notifications@github.com>
发送时间: 2018年1月3日 19:00
收件人: zerotier/ZeroTierNAS
抄送: hsfq; Comment
主题: Re: [zerotier/ZeroTierNAS] 88f6281 Downlaod non existant (#12)


Working to fix this. At the worst case I'll just re-package one of the 1.2.4 binaries I cross compiled a while back for this arch. Can you both tell me the exact models and exact DSM versions?

Thanks.

―

**Comment by joseph-henry** [comment]
If you want to use this with DSM 5.X it looks like we'll need to use a different packaging system:

https://developer.synology.com/developer-guide/create_package/index.html

Since we already have a binary for 88F6281/2, we can skip steps 1, 2, and 4.

We would likely use this:

https://github.com/SynologyOpenSource/pkgscripts

I can look into it tomorrow.

**Comment by lucktu** [comment]
Yes, it's very kind of you , Thanks!!

I don't update the dsm 5.2  for my timebackup forever

________________________________
发件人: Joseph Henry <notifications@github.com>
发送时间: 2018年1月4日 3:25
收件人: zerotier/ZeroTierNAS
抄送: hsfq; Comment
主题: Re: [zerotier/ZeroTierNAS] 88f6281 Download non existent (#12)


If you want to use this with DSM 5.X it looks like we'll need to use a different packaging system:

https://developer.synology.com/developer-guide/create_package/index.html

Create Package | Synology DSM6.0 Developer Guide<https://developer.synology.com/developer-guide/create_package/index.html>
developer.synology.com
Create Package. In this section, we will explain how to create a Synology Package using Package Toolkit. If you want to build a Synology Package without using Package ...



Since we already have a binary for 88F6281/2, we can skip steps 1, 2, and 4.

We would likely use this:

https://github.com/SynologyOpenSource/pkgscripts

[https://avatars2.githubusercontent.com/u/13974755?s=400&v=4]<https://github.com/SynologyOpenSource/pkgscripts>

GitHub - SynologyOpenSource/pkgscripts: Synology DSM5.x ...<https://github.com/SynologyOpenSource/pkgscripts>
github.com
You can download and set up pre-built environments by using **EnvDeploy **as follows. Use -v to specify DSM version and -p to specify desired platform. If -p is not ...



I can look into it tomorrow.

―

**Comment by meoso** [comment]
@joseph-henry  to answer your request: `DS410j` with `DSM 5.2-5967 Update 6`
Thank you for your efforts!

**Comment by joseph-henry** [comment]
We won't be officially supporting DSM versions older than 6.0, unfortunately it's too time consuming to build packages for these older versions. Synology's packaging system is not well documented and contains too many documentation errors.

But! Try extracting the binary from the package you linked:
https://download.zerotier.com/dist/zerotier-1.2.4-syn-88f6281-6.1.spk

It's actually a `tar.gz` file in disguise. Open that, and then open the `package.tar.gz` file inside that. You'll find a `zerotier-one` binary. Copy that to your NAS and try to run it as root. 

If it runs without crashing, then I can give you further instructions.

**Comment by meoso** [comment]
@joseph-henry it runs, but did not pickup my old settings.  tried to `zerotier-one -q join ###########` but reports `401 join {}`

```
zerotier-one -q info
401 info {}
```

everything i throw at it results `401 command {}`

**Comment by lucktu** [comment]
@joseph-henry  OK, no crashing
ABC> ./zerotier-one
connect: No route to host
connect: No route to host
connect: No route to host
connect: Connection timed out
connect: Connection timed out
... ...

**Comment by lucktu** [comment]
@joseph-henry 
But then ? give me further instructions?
I will use it in other place.
Thanks!

**Comment by joseph-henry** [comment]
Sure, here is the script that the DSM 6.0 package executes during install: https://github.com/zerotier/ZeroTierNAS/blob/master/Synology/package/zerotier/spk/scripts/postinst

I've taken out the DSM 6.0-specific parts. This should get you most of the way. But you'll probably need to find a way to manually register this as a service so it starts up automatically.

```
#!/bin/bash
mkdir -p /usr/local/zerotier/bin
mkdir -p /var/lib/zerotier-one
chmod 755 zerotier-one
ln -s zerotier-one zerotier-cli
cp zerotier-one /usr/local/zerotier/bin
cp zerotier-cli /usr/local/zerotier/bin

# Create the necessary TUN (might need to do this every reboot)
SERVICE="zerotier"
ZEROTIER_MODULE="tun.ko"
BIN_SYNOMODULETOOL="/usr/syno/bin/synomoduletool"

# Make device if not present (not devfs)
echo "$(date) INSTALL: creating /dev/net/tun device" >> /opt/var/log/zerotier-one.log ;
if ( [ ! -c /dev/net/tun ] ) then
		# Make /dev/net directory if needed
		if ( [ ! -d /dev/net ] ) then
    		mkdir -m 755 /dev/net
		fi
		mknod /dev/net/tun c 10 200
fi

# Load TUN kernel module
echo "$(date) INSTALL: loading TUN kernel module" >> /opt/var/log/zerotier-one.log ;
if [ -x ${BIN_SYNOMODULETOOL} ]; then
	$BIN_SYNOMODULETOOL --insmod $SERVICE ${ZEROTIER_MODULE}
else
	/sbin/insmod /lib/modules/${ZEROTIER_MODULE}
fi

exit 0
```

Additionally, we use the `zerotier-watchdog.sh` to make sure Synology doesn't delete routes, or uninstall the TUN driver. https://github.com/zerotier/ZeroTierNAS/blob/master/Synology/package/zerotier/app/bin/zerotier-watchdog.sh

**Comment by lucktu** [comment]
OK, This is what I want to get.
Thank you!!

**Comment by meoso** [comment]
got it working; however i see multiple `zerotier-one -d` going up and down when i `ps | grep zero`

and if i run the watchdog script with `start` it will also introduce another daemon (both script and binary). 

i'll have to investigate this further, i'll revert to 1.1.14 until i have significant time.

**Comment by joseph-henry** [comment]
The only suggestion I can make is to try to cannibalize the latest `watchdog` and `start-stop-status` scripts. As far as I can tell they work very well on DSM 6.1+.

The following commit fixed most issues for people: https://github.com/zerotier/ZeroTierNAS/commit/a2e87ff2c09c3da9ac3efdb6803bde29b1a64082

**Comment by meoso** [comment]
For records.

With the binary `zerotier-one` from from https://download.zerotier.com/dist/zerotier-1.2.8r0-syn-88f6281-6.1.spk

```
root@DiskStation:/ # ls -l /opt/bin/zerotier-*
lrwxrwxrwx    1 root     root            12 Apr 26  2017 /opt/bin/zerotier-cli -> zerotier-one
lrwxrwxrwx    1 root     root            12 Apr 26  2017 /opt/bin/zerotier-idtool -> zerotier-one
-rwxr-xr-x    1 root     root       1882200 Jul 19 13:16 /opt/bin/zerotier-one
```

Coupled with my cronjobbed "service" script (https://gist.github.com/meoso/b25bd410c8a54a1a013f0cc2d72b12ee), I have v1.2.x running on my Synology 410j (Marvell Kirkwood 88F6281 / ARMv5).  I just have it `start` every 5 minutes, which will just bypass if already running.

I only have one issue, bu it does not affect usage:
```
ERROR: unable to add ip addresses to ifcfg
```

If you know why and how to avoid, i'd appreciate feedback.

Thank you!


p.s. for others looking for solutions, this may help: https://gist.github.com/DraTeots/c2519efdd3774b6044e2b2d6e977f24d

**Comment by hawkinchina** [comment]
@joseph-henry  hi, it can work for DS3615xs 5.2  by   the same way of  zerotier_bromolow-6.0.2_1.4.0-0.spk    or    zerotier_bromolow-6.1_1.4.0-0.spk   ?

**Comment by eric4gith** [comment]
I have a DS1010+ with DSM 5.2. "zerotier-1.2.4-noarch.spk" is the only working package I found working. It has a problem. That is, every time the NAS got reboot, or shutdown then start again, ZeroTier One fails to join ZeroTier network. I have to manually uninstall ZeroTier package, install again, and joint ZeroTier network through CLI.
Is there a solution or newer package that available for DS1010+ with DSM 5.2?

**Comment by eric4gith** [comment]
I have solved my problem. Post it for record and reference.
I followed the instructions for creating docker on Synology NAS: https://docs.zerotier.com/devices/synology/
I only did the "Create a persistent TUN" part, no Docker container. Still using "zerotier-1.2.4-noarch.spk". ZeroTier One can join ZTN automatically now.
